tests/cpp_unit_tests/test_tap_position_optimizer.cpp

```diff
@@ -2,42 +2,127 @@
+                                 {5, 50e3},  {6, 10e3}, {7, 10e3}, {8, 10e3}, {9, 10e3}};
+    main_core::add_component<Node>(state, nodes.begin(), nodes.end(), 50.0);
+
+    std::vector<TransformerInput> transformers{
+        get_transformer(11, 0, 1, 150e3, 10e3), get_transformer(12, 0, 1, 150e3, 10e3),
+        get_transformer(13, 5, 7, 50e3, 10e3), get_transformer(14, 2, 3, 10e3, 10e3),
+        get_transformer(15, 8, 9, 10e3, 10e3)};
+    main_core::add_component<Transformer>(state, transformers.begin(), transformers.end(), 50.0);
+
+    std::vector<LineInput> lines{get_line_input(16, 1, 2), get_line_input(17, 4, 6), get_line_input(18, 7, 8),
+                                 get_line_input(19, 3, 6)};
+    main_core::add_component<Line>(state, lines.begin(), lines.end(), 50.0);
+
+    std::vector<LinkInput> links{{20, 2, 1, 1, 1}, {21, 6, 4, 1, 1}, {22, 8, 7, 1, 1}, {23, 9, 3, 1, 1}};
+    main_core::add_component<Link>(state, links.begin(), links.end(), 50.0);
+
+    std::vector<SourceInput> sources{{24, 0, 1, 1.0, 0, nan, nan, nan}};
+    main_core::add_component<Source>(state, sources.begin(), sources.end(), 50.0);
+
+    std::vector<TransformerTapRegulatorInput> regulators{{25, 11, 1, ControlSide::from, nan, nan, nan, nan},
+                                                         {26, 12, 1, ControlSide::from, nan, nan, nan, nan},
+                                                         {27, 13, 1, ControlSide::from, nan, nan, nan, nan},
+                                                         {28, 14, 1, ControlSide::from, nan, nan, nan, nan},
+                                                         {29, 15, 1, ControlSide::from, nan, nan, nan, nan}};
+    main_core::add_component<TransformerTapRegulator>(state, regulators.begin(), regulators.end(), 50.0);
+
+    state.components.set_construction_complete();
 
     // Subcases
     SUBCASE("Building the graph") {
-        // ToDo: graph creation
+        // reference graph creation
+        std::vector<std::pair<Idx, Idx>> edge_array = {{0, 1}, {0, 2}, {2, 3}};
+
+        std::vector<pgm_tap::TrafoGraphEdge> edge_prop{{{0, 1}, 1}, {{0, 1}, 1}, {{0, 1}, 1}};
+        pgm_tap::TransformerGraph expected_graph{boost::edges_are_unsorted_multi_pass, edge_array.cbegin(),
+                                                 edge_array.cend(), edge_prop.cbegin(), 10};
+        pgm_tap::TransformerGraph actual_graph = pgm_tap::build_transformer_graph(state);
+
+        std::vector<pgm_tap::TrafoGraphVertex> expected_vertex_props{{true},  {false}, {false}, {false}, {false},
+                                                                     {false}, {false}, {false}, {false}, {false}};
+
+        boost::graph_traits<pgm_tap::TransformerGraph>::vertex_iterator vi, vi_end;
+        for (boost::tie(vi, vi_end) = vertices(actual_graph); vi != vi_end; ++vi) {
+            actual_graph[*vi].is_source = expected_vertex_props[*vi].is_source;
+        }
```

Comment:
Why is this necessary? the `// Mark sources` section inside the `build_transformer_graph` function is already doing that, no?